rpcserver.go

```diff
@@ -5397,3 +5404,149 @@ func (r *rpcServer) BakeMacaroon(ctx context.Context,
 
 	return resp, nil
 }
+
+// ChannelRevenue returns the revenue for a channel over a period of time.
+// To limit the size of query required to calculate revenue, channel open
+// an close times are looked up by block height, and the user provided period
+// is bounded by these values; no revenue generating events can occur before
+// channel open or after channel close.
+func (r *rpcServer) ChannelRevenue(ctx context.Context,
+	req *lnrpc.ChannelRevenueRequest) (*lnrpc.ChannelRevenueResponse, error) {
+
+	// Get the outpoint from the request.
+	txid, err := GetChanPointFundingTxid(req.GetChannelPoint())
+	if err != nil {
+		return nil, err
+	}
+	outpoint := wire.NewOutPoint(txid, req.ChannelPoint.OutputIndex)
+
+	var (
+		channelClosed bool
+		closeHeight   uint32
+		shortChanID   lnwire.ShortChannelID
+	)
+
+	// Attempt to lookup the channel, first checking open channels and then
+	// closed if it is not found. We will save the short channel ID (containing
+	// open height) and close height (if the channel is closed) so that we can
+	// obtain channel open/close times from the block height. These times are
+	// needed to reduce to size of the forwarder log query that get revenue
+	// requires.
+
+	channel, err := r.server.chanDB.FetchChannel(*outpoint)
+	// If the channel is not found, set channelClosed to true so that we know
+	// to look for a closed channel.
+	if err == channeldb.ErrChannelNotFound {
+		channelClosed = true
+	} else if err != nil {
+		return nil, err
+	}
+
+	// If the channel was found, set the short channel id from the channel.
+	if channel != nil {
+		shortChanID = channel.ShortChannelID
+	}
+
+	// If we did not find an open channel with the outpoint provided, check
+	// whether the channel is already closed.
+	if channelClosed {
+		channel, err := r.server.chanDB.FetchClosedChannel(outpoint)
+		// If a closed channel is not found either, return a generic channel
+		// not found error because the channel is unknown to lnd.
+		if err == channeldb.ErrClosedChannelNotFound {
+			return nil, fmt.Errorf("channel: %v not founnd", outpoint)
+		} else if err != nil {
+			return nil, err
+		}
+
+		// Set the channel ID and close height from the close summary.
+		shortChanID = channel.ShortChanID
+		closeHeight = channel.CloseHeight
+	}
+
+	// Set start time to the user specified value. If no value is provided,
+	// start time will be set to the unix epoch.
+	startTime := time.Unix(int64(req.StartTime), 0)
+
+	// Set end time from the user specified time. If no value was provided,
+	// end time will be set to the present.
+	endTime := time.Unix(int64(req.EndTime), 0)
+	if req.EndTime == 0 {
+		endTime = time.Now()
+	}
+
+	// Get open time for the channel from block height.
```

Comment:
Many items below this block can likely now be simplified after the latest set of offline discussions.